File: model-optimizer/mo/front/common/partial_infer/slice.py
"""
 Copyright (c) 2018-2019 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""

import numpy as np

from mo.utils.error import Error


def tf_strided_slice_infer(node):
    if node.in_node(1).value is None or node.in_node(2).value is None:
        raise Error('Strided slice layer supports only constant begin and end inputs')
    begin_id = node.in_node(1).value.copy()
    end_id = node.in_node(2).value.copy()
    if len(node.in_nodes()) > 3:
        if node.in_node(3).value is None:
            raise Error('Strided slice layer supports only constant stride input')
        stride = node.in_node(3).value
    else:
        stride = []

    shape = node.in_node(0).shape

    if shape is None or any([x < 0 for x in shape]):
        return

    convert_negative_indices(begin_id, shape)
    convert_negative_indices(end_id, shape)

    slice_idx = []
    dims = np.amax(np.array([len(begin_id), len(end_id), len(stride),
                             len(node.shrink_axis_mask), len(node.new_axis_mask), len(node.ellipsis_mask),
                             len(node.begin_mask), len(node.end_mask)]))

    # make mask correct length
    def extend_mask(in_mask, fin_len, zeros=True):
        mask = list(in_mask)
        if len(mask) < fin_len:
            if zeros:
                mask.extend(np.zeros(dims-len(mask), dtype=np.int32))
            else:
                mask.extend(np.ones(dims-len(mask), dtype=np.int32))
        return np.array(mask, dtype=np.int32)

    for mask in {'new_axis_mask', 'shrink_axis_mask', 'ellipsis_mask'}:
        node[mask] = extend_mask(node[mask], dims)
    node.begin_mask = extend_mask(node.begin_mask, dims, False)
    node.end_mask = extend_mask(node.end_mask, dims, False)

    old_idx = 0
    ellips_ext = 0
    id_em = 0
    for idx in range(dims):
        if node.new_axis_mask[idx]:
            slice_idx.append(np.newaxis)
        elif node.ellipsis_mask[idx]:
            ellips_ext = len(shape) - (dims - np.count_nonzero(node.new_axis_mask) - 1)
            id_em = idx
            for i in range(0, ellips_ext):
                slice_idx.append(slice(0, shape[old_idx], 1))
                old_idx = old_idx + 1
        else:
            s = stride[idx] if len(stride) > idx else 1
            def_beg = 0 if s > 0 else -1
            def_end = shape[old_idx] if s > 0 else -shape[old_idx]-1
            l = begin_id[idx] if node.begin_mask[idx] and idx < len(begin_id) else def_beg
            r = end_id[idx] if node.end_mask[idx] and idx < len(end_id) else def_end

            # Check shrink_axis_mask
            if node.shrink_axis_mask[idx] and idx < len(shape):
                slice_idx.append(slice(l, l+1, s))
            else:
                slice_idx.append(slice(l, r, s))
            old_idx = old_idx + 1

    value = node.in_node(0).value if node.in_node(0).value is not None else np.zeros(shape)
    # fix for the warning: "FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated use
    # `arr[tuple(seq)]` instead of `arr[seq]`"
    value = value[tuple(slice_idx)]

    for idx, flag in reversed(list(enumerate(node.shrink_axis_mask))):
        if flag:
            if ellips_ext > 0 and idx > id_em:
                idx = idx + ellips_ext - 1
            try:
                value = np.squeeze(value, idx)
            except ValueError:
                # ignore this error
                continue

    node['slices'] = np.array(slice_idx)
    for attr in ('shrink_axis_mask', 'new_axis_mask', 'ellipsis_mask', 'begin_mask', 'end_mask'):
        node[attr] = np.array(node[attr], dtype=np.int32)

    node.out_node().value = np.array(value) if node.in_node(0).value is not None else None
    node.out_node().shape = np.array(value.shape, dtype=np.int64)

    # change precision to I32 for begin, end, stride inputs
    for i in range(1, len(node.in_nodes())):
        inp = node.in_node(i)
        inp["force_precision"] = "I32"


def convert_negative_indices(indices: np.array, shape: np.array):
    for ind, value in enumerate(indices):
        if value < 0:
            indices[ind] += shape[ind]


def caffe_slice_infer(node):
    """
    Slices an input layer to multiple output layers along a given dimension
    with given slice indices
    Parameters
    ----------
    node

    """
    top_shape = node.in_node(0).shape
    slice_axis = node.axis
    bottom_slice_axis = node.in_node(0).shape[node.axis]
    if len(node.slice_point) == 0:
        new_shape = np.array(top_shape, dtype=np.int64)
        new_shape[slice_axis] = bottom_slice_axis / len(node.out_nodes())
        for i in range(0, len(node.out_nodes())):
            node.out_node(i).shape = new_shape
        return

    assert (len(node.slice_point) == len(node.out_nodes()) - 1)
    prev = 0
    slices = []
    for slice_point in node.slice_point:
        if slice_point <= prev:
            raise Error(
                'Check failed for the layer {}. Slice points should be ordered in increasing manner. '.format(node.id) +
                'Current slice point {} is not greater than the previous slice point {}. '.format(slice_point, prev) +
                'Please verify your model correctness')
        slices.append(slice_point - prev)
        prev = slice_point

    slices.append(bottom_slice_axis - prev)
    if sum(slices) != bottom_slice_axis:
        raise Error(
            'Check failed for the layer {}. Sum of slices points {} does not equal '.format(node.id, sum(slices)) +
            'to the value of input blob shape by the given slice axis {}'.format(bottom_slice_axis))
    for i in range(len(node.out_nodes())):
        new_shape = np.array(top_shape, dtype=np.int64)
        new_shape[slice_axis] = slices[i]
        node.out_node(i).shape = new_shape


def mxnet_slice_axis_infer(node):
    in_shape = node.in_node(0).shape
    slice_axis = node.axis

    new_shape = np.array(in_shape, dtype=np.int64)
    new_shape[slice_axis] = new_shape[slice_axis] / len(node.out_nodes())

    axis_size = in_shape[slice_axis]
    if node.offset < 0:
        node.offset += axis_size

    if not node.dim:
        node.dim = axis_size
    elif node.dim < 0:
        node.dim += axis_size

    input_dim = in_shape.size
    node.dim = (node.dim - node.offset)
    if node.dim > in_shape[slice_axis]:
        raise Error(
            '{0} node dimension value is bigger than the corresponding value in the input shape {1}. ' +
            '\nIn particular {2} is bigger than {3}. The Model Optimizer does not support this case. ' +
            '\nTo overcome, try to edit the original model "end" property of the {0} layer.',
            node.name, ','.join(str(i) for i in in_shape), str(node.dim), str(in_shape[slice_axis])
        )

    for i in range(0, input_dim):
        if i == slice_axis:
            new_shape[i] = node.dim
        else:
            new_shape[i] = in_shape[i]

    for i in range(0, len(node.out_nodes())):
        node.out_node(i)['shape'] = new_shape

File: ngraph/src/ngraph/op/slice.cpp
//*****************************************************************************
// Copyright 2017-2020 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

#include "ngraph/op/slice.hpp"

using namespace std;
using namespace ngraph;

constexpr NodeTypeInfo op::Slice::type_info;

op::Slice::Slice(const Output<Node>& arg,
                 const Coordinate& lower_bounds,
                 const Coordinate& upper_bounds,
                 const Strides& strides)
    : Op({arg})
    , m_lower_bounds(lower_bounds)
    , m_upper_bounds(upper_bounds)
    , m_strides(strides)
{
    constructor_validate_and_infer_types();
}

op::Slice::Slice(const Output<Node>& arg,
                 const Coordinate& lower_bounds,
                 const Coordinate& upper_bounds)
    : Op({arg})
    , m_lower_bounds(lower_bounds)
    , m_upper_bounds(upper_bounds)
    , m_strides(Strides())
{
    constructor_validate_and_infer_types();
}

void op::Slice::validate_and_infer_types()
{
    // An empty stride vector with lower_bounds/upper_bounds filled in means that we need to
    // construct the default value.
    if (m_strides.size() == 0)
    {
        m_strides = Strides(m_lower_bounds.size(), 1);
    }

    NODE_VALIDATION_CHECK(this,
                          m_lower_bounds.size() == m_upper_bounds.size() &&
                              m_lower_bounds.size() == m_strides.size(),
                          "Ranks of lower bounds (",
                          m_lower_bounds,
                          "), upper bounds (",
                          m_upper_bounds,
                          ") and strides (",
                          m_strides,
                          ") do not match.");

    size_t output_rank = m_upper_bounds.size();

    for (size_t i = 0; i < output_rank; i++)
    {
        NODE_VALIDATION_CHECK(this,
                              m_lower_bounds[i] <= m_upper_bounds[i],
                              "Lower bound for slice is greater than upper bound at axis ",
                              i,
                              " (lower bounds: ",
                              m_lower_bounds,
                              ", upper bounds: ",
                              m_upper_bounds,
                              ").");

        NODE_VALIDATION_CHECK(this,
                              m_strides[i] != 0,
                              "Stride for slice is zero at axis ",
                              i,
                              " (strides: ",
                              m_strides,
                              ").");
    }

    const PartialShape& input_shape = get_input_partial_shape(0);
    Dimension input_rank = input_shape.rank();

    NODE_VALIDATION_CHECK(this,
                          input_rank.is_dynamic() || input_rank.get_length() == output_rank,
                          "Input rank does not match the rank of the lower bounds (",
                          m_lower_bounds,
                          "), upper bounds (",
                          m_upper_bounds,
                          "), and strides (",
                          m_strides,
                          ").");

    std::vector<Dimension> result_dims(output_rank);

    for (size_t i = 0; i < output_rank; i++)
    {
        NODE_VALIDATION_CHECK(this,
                              input_rank.is_dynamic() || input_shape[i].is_dynamic() ||
                                  m_upper_bounds[i] <= input_shape[i].get_length(),
                              "Upper bound for slice at axis ",
                              i,
                              " is out of range ",
                              "(upper bounds: ",
                              m_upper_bounds,
                              ", argument shape: ",
                              input_shape,
                              ").");

        size_t result_axis_size = m_upper_bounds[i] - m_lower_bounds[i];
        result_axis_size =
            result_axis_size / m_strides[i] + ((result_axis_size % m_strides[i] == 0) ? 0 : 1);
        result_dims[i] = result_axis_size;
    }

    set_output_type(0, get_input_element_type(0), PartialShape{result_dims});
}

shared_ptr<Node> op::Slice::clone_with_new_inputs(const OutputVector& new_args) const
{
    check_new_args_count(this, new_args);
    return make_shared<Slice>(new_args.at(0), m_lower_bounds, m_upper_bounds, m_strides);
}

void op::Slice::generate_adjoints(autodiff::Adjoints& adjoints, const OutputVector& deltas)
{
    auto delta = deltas.at(0);

    auto x = input_value(0);

    adjoints.add_delta_to_slice(x, delta, m_lower_bounds, m_upper_bounds, m_strides);
}
